### PR TITLE
Better implementation of native menu bar on Linux

### LINKED ISSUE #18572: Crash on quit (Linux with enabled global menu bar)
### Issue type

Crash or freeze

### Bug description

MuseScore 4.1 crashes on quit when the platform menu bar on Linux is enabled.

### Steps to reproduce

1. Enable global menu bar in your desktop environment.
2. Open MuseScore.
3. Close MuseScore.

### Screenshots/Screen recordings

_No response_

### MuseScore Version

4.1

### Regression

Yes, this used to work in a previous version of MuseScore 4.x

### Operating system

Kubuntu 22.04 (KDE Plasma 5.24.7)

### Additional context

```
#0  0x00007ffff5b5afb4 in  () at /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#1  0x00007ffff5d785ec in QObjectPrivate::connectImpl(QObject const*, int, QObject const*, void**, QtPrivate::QSlotObjectBase*, Qt::ConnectionType, int const*, QMetaObject const*) ()
    at /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#2  0x00007ffff5d78b15 in QObject::connectImpl(QObject const*, void**, QObject const*, void**, QtPrivate::QSlotObjectBase*, Qt::ConnectionType, int const*, QMetaObject const*) ()
    at /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#3  0x00007fff7eb6bbfc in  ()
    at /usr/lib/x86_64-linux-gnu/qt5/qml/Qt/labs/platform/libqtlabsplatformplugin.so
#4  0x00007fff7eb6d1c0 in  ()
    at /usr/lib/x86_64-linux-gnu/qt5/qml/Qt/labs/platform/libqtlabsplatformplugin.so
#5  0x00007fff7eb6d23d in  ()
    at /usr/lib/x86_64-linux-gnu/qt5/qml/Qt/labs/platform/libqtlabsplatformplugin.so
#6  0x00007fff7eb5f457 in  () at /usr/lib/x86_64-linux-gnu/qt5/qml/Qt/labs/platform/libqtlabsplatformplugin.so
#7  0x00007ffff5d74a6e in QObjectPrivate::deleteChildren() () at /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#8  0x00007ffff5d7f924 in QObject::~QObject() () at /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#9  0x00007fff7eb5f4c7 in  () at /usr/lib/x86_64-linux-gnu/qt5/qml/Qt/labs/platform/libqtlabsplatformplugin.so
#10 0x00007ffff5d74a6e in QObjectPrivate::deleteChildren() () at /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#11 0x00007ffff5d7f924 in QObject::~QObject() () at /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#12 0x00007ffff7c63859 in  () at /usr/lib/x86_64-linux-gnu/libQt5Quick.so.5
#13 0x00007ffff5d74a6e in QObjectPrivate::deleteChildren() () at /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#14 0x00007ffff5d7f924 in QObject::~QObject() () at /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#15 0x00007ffff7c64789 in  () at /usr/lib/x86_64-linux-gnu/libQt5Quick.so.5
#16 0x00007ffff5d74a6e in QObjectPrivate::deleteChildren() () at /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#17 0x00007ffff5d7f924 in QObject::~QObject() () at /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#18 0x00007fff8d713381 in  () at /usr/lib/x86_64-linux-gnu/qt5/qml/QtQuick/Templates.2/libqtquicktemplates2plugin.so
#19 0x00007ffff76e29ce in QQmlApplicationEnginePrivate::cleanUp() () at /usr/lib/x86_64-linux-gnu/libQt5Qml.so.5
#20 0x00007ffff76e2a1d in QQmlApplicationEngine::~QQmlApplicationEngine() () at /usr/lib/x86_64-linux-gnu/libQt5Qml.so.5
#21 0x00007ffff76e2a3d in QQmlApplicationEngine::~QQmlApplicationEngine() () at /usr/lib/x86_64-linux-gnu/libQt5Qml.so.5
#22 0x0000555556b8f9c7 in mu::ui::UiEngine::quit() (this=0x5555596eecc0 <mu::ui::UiEngine::instance()::e>) at /home/adazem009/tmp/MuseScore/src/framework/ui/internal/uiengine.cpp:66
#23 0x0000555555784965 in mu::app::App::run(int, char**) (this=0x7fffffffd820, argc=1, argv=0x7fffffffdd38) at /home/adazem009/tmp/MuseScore/src/app/app.cpp:353
#24 0x000055555576c1ae in main(int, char**) (argc=1, argv=0x7fffffffdd38) at /home/adazem009/tmp/MuseScore/src/app/main.cpp:341
```
Might be related to #16456 because Linux uses the same platform menu bar implementation as macOS.

### PR DESCRIPTION
**Pull request by adazem009**
Resolves: #18572
Resolves: #16094

This PR includes these fixes:

- I disabled loading menu subitems when a menu is created before because it caused a crash on launch. Loading subitems after adding the menu to the menu bar fixes the crash.
- Re-creating menu items when a menu is about to show causes a crash on quit and flickering glitches when opening menus. I made it create the menu items only when the app starts and when the `subitems` property changes (see the connections in the `load()` function). The titles, checked, enabled, etc. properties are updated when a menu is about to show.

Note: I didn't test this on macOS. The original menu bar implementation should be used there.

- [x] I signed the [CLA](https://musescore.org/en/cla)
- [x] The title of the PR describes the problem it addresses
- [x] Each commit's message describes its purpose and effects, and references the issue it resolves
- [x] If changes are extensive, there is a sequence of easily reviewable commits
- [x] The code in the PR follows [the coding rules](https://github.com/musescore/MuseScore/wiki/CodeGuidelines)
- [x] There are no unnecessary changes
- [x] The code compiles and runs on my machine, preferably after each commit individually
- [ ] I created a unit test or vtest to verify the changes I made (if applicable)

### CODE REVIEW

**Comment by Eism** [comment]
@adazem009 
It would be great to figure out why this started crashing. After a quick code review of your previous PR #15826 and the current master branch, I don't see any differences.
Can you reproduce the bug on your previous PR? (you'll need to build that branch)

File: src/appshell/qml/platform/PlatformMenuBar.qml [hunk 1]
```diff
@@ -104,6 +100,8 @@ Item {
             property var subitems: []
 
             onAboutToShow: {
+                for(var j in items)
+                    items[j].destroy()
```
Comment (Eism) [inline src/appshell/qml/platform/PlatformMenuBar.qml:104]: run: error: qrc:/qml/platform/PlatformMenuBar.qml:106: Error: Invalid attempt to destroy() an indestructible object

macOS

**Comment by adazem009** [comment]
> @adazem009 It would be great to figure out why this started crashing. After a quick code review of your previous PR #15826 and the current master branch, I don't see any differences. Can you reproduce the bug on your previous PR? (you'll need to build that branch)

Yes, that was the first thing I had done before I came up with this fix on the master branch.

File: src/appshell/qml/platform/PlatformMenuBar.qml (hunk 1 above)
Comment (adazem009) [reply src/appshell/qml/platform/PlatformMenuBar.qml:104]: That's strange, it fixes the crash on Linux for some reason. I'll have a look at the logs...

**Comment by adazem009** [comment]
@Eism I really think this is a Qt bug. I should test it with a newer Qt 5.15 release instead of doing weird workarounds.
Anyway, I've found a different workaround (in the latest commit). The crash doesn't occur when I set the `visible` property of root menus to `false` when they're about to hide.

This crash was happening at the time of merging #15826, but I just didn't notice it.

By the way, I think the menu bar needs a different implementation. What's the point of re-creating menu items every time when they're shown to the user (the `aboutToShow` signal)? Do the menus ever change (except the `enabled` property)? Maybe we should ask @RomanPudashkin about this. See this commit: ec8c194a8ea6a7a51b97e3673c8e141f45ff6577

**Comment by cbjeukendrup** [comment]
Regarding using a newer Qt version, this might be helpful, from another PR that was also about a Qt bug: https://github.com/musescore/MuseScore/pull/17997#issuecomment-1616618937

**Comment by MarcSabatella** [comment]
I don't know if this it actually relevant, but as long as you're looking at how menus are created on Linux, it sure would be nice if the new approach somehow managed to work on Wayland as per https://github.com/musescore/MuseScore/issues/15285.

**Comment by adazem009** [comment]
> I don't know if this it actually relevant, but as long as you're looking at how menus are created on Linux, it sure would be nice if the new approach somehow managed to work on Wayland as per #15285 (see in particular the comment about global menus in [#15285 (comment)](https://github.com/musescore/MuseScore/issues/15285#issuecomment-1368494118))

That issue is related to the non-native menu bar. I tried to launch MuseScore under Wayland but it doesn't start (see #18598).
However, it works when I build it from source and I can reproduce the issue you've linked. I've also noticed that the native menu bar doesn't even show, so there isn't any menu bar when global menu bar is enabled in the desktop environment. But this should go into a new issue.

**Comment by cbjeukendrup** [comment]
> @cbjeukendrup Do you think it's better to avoid re-creating menu items every time a menu is shown to the user?

I'll defer this question to @RomanPudashkin and @Eism, since I'm not sure why this was done in the first place; but if it works without it, I can imagine that would be preferable because more efficient.

**Comment by cbjeukendrup** [comment]
I tried out the latest version on macOS and found some strange things: "About MuseScore" and "About Qt" now open the "About MusicXML" dialog, and the "Preferences" item calls the "revert-factory" action. This seems related to the [`menuItem.role`](https://doc.qt.io/qt-5/qml-qt-labs-platform-menuitem.html#role-prop) property, since all the affected menu items do something with this property.
Since this property influences in which menu a menu item appears, it might be necessary to set this very early, or for example to make sure that all MenuItems have this property set before the menus containing them are attached to the menu bar.

**Comment by cbjeukendrup** [comment]
I found another thing: directly after launching MuseScore, these affected items initially don't appear on macOS. 
On macOS, they are supposed to appear in the "MuseScore" menu, which is added by macOS as the leftmost menu. We simply add them to the "Edit" and "Help" menus, where they are supposed to be on other platforms, but because of the `role` property, they will be moved automatically from those menus to the MuseScore menu.
So, initially, after launching the app, these items don't appear in the MuseScore menu. But once you've opened the "Edit" menu, the "Preferences" item is added to the MuseScore menu (but does not do the thing it's supposed to do). And once you've opened the "Help" menu, the "About" items appear in the MuseScore menu too. Maybe that's useful info.

If it turns out to be difficult to solve, we can decide to make two independent native menu bars, one for Mac and one for Linux.

**Comment by adazem009** [comment]
@cbjeukendrup This doesn't happen on Linux, but I think it's caused by how the menus are created. When you launch MuseScore, the `load()` function is called on each menu. This creates empty menu items and submenus (and items in the submenus). When a menu is about to show (`onAboutToShow`), the `update()` function is called on the menu which sets the properties of each menu item.
Maybe calling `update()` on each menu when MuseScore starts would fix this, but I'll need some help with this because I don't have a Mac and it'd take ages to launch MuseScore on my Hackintosh VM :)

**Comment by cbjeukendrup** [comment]
I see. Then the best option is probably that I take a look at these issues myself :)

**Comment by DmitryArefiev** [comment]
@adazem009 Hi! Can you rebase please?

**Comment by adazem009** [comment]
> @adazem009 Hi! Can you rebase please?

Done

**Comment by DmitryArefiev** [comment]
Thanks! 

I installed KDE plasma on my Ubuntu finally and able to reproduce crash from #18572 in master

Tested #18572 on Ubuntu 22.04 with KDE Plasma 5.24.7 - **FIXED**

Tested #16094 on Ubuntu 22.04 with KDE Plasma 5.24.7 - probably **FIXED**  see https://github.com/musescore/MuseScore/issues/16094#issuecomment-1816324869

**Review by cbjeukendrup — APPROVED** [review]

**Pull request merged**